src/RecurringPurchase.sol

```diff
@@ -0,0 +1,270 @@
+import {ISwapRouter} from "v3-periphery/interfaces/ISwapRouter.sol";
+
+import {QuarkWallet} from "quark-core/src/QuarkWallet.sol";
+import {QuarkScript} from "quark-core/src/QuarkScript.sol";
+
+/**
+ * @title Recurring Purchase Script
+ * @notice Quark script that performs a swap on a regular interval.
+ * @author Legend Labs, Inc.
+ */
+contract RecurringPurchase is QuarkScript {
+    using SafeERC20 for IERC20;
+
+    error BadPrice();
+    error InvalidInput();
+    error PurchaseConditionNotMet();
+
+    /// @notice Emitted when a swap is executed
+    event SwapExecuted(
+        address indexed sender,
+        address indexed recipient,
+        address tokenIn,
+        address tokenOut,
+        uint256 amountIn,
+        uint256 amountOut,
+        bytes path
+    );
+
+    /// @notice The base slippage factor where `1e18` represents 100% slippage tolerance
+    uint256 public constant BASE_SLIPPAGE_FACTOR = 1e18;
+
+    /**
+     * @dev Note: This script uses the following storage layout in the QuarkStateManager:
+     *         mapping(bytes32 hashedPurchaseConfig => uint256 nextPurchaseTime)
+     *             where hashedPurchaseConfig = keccak256(PurchaseConfig)
+     */
+
+    /// @notice Parameters for a recurring purchase order
+    struct PurchaseConfig {
+        uint256 interval;
+        SwapParams swapParams;
+        SlippageParams slippageParams;
+    }
+
+    /// @notice Parameters for performing a swap
+    struct SwapParams {
+        address uniswapRouter;
+        address recipient;
+        address tokenIn;
+        address tokenOut;
+        /// @dev The amount for tokenIn if exact in; the amount for tokenOut if exact out
+        uint256 amount;
+        /// @dev False for exact in; true for exact out
+        bool isExactOut;
+        uint256 deadline;
+        bytes path;
+    }
+
+    /// @notice Parameters for controlling slippage in a swap operation
+    struct SlippageParams {
+        /// @dev Maximum acceptable slippage, expressed as a percentage where 100% = 1e18
+        uint256 maxSlippage;
+        /// @dev Price feed addresses for determining market exchange rates between token pairs
+        /// Example: For SUSHI -> SNX swap, use [SUSHI/ETH feed, SNX/ETH feed]
+        address[] priceFeeds;
+        /// @dev Flags indicating whether each corresponding price feed should be inverted
+        /// Example: For USDC -> ETH swap, use [true] with [ETH/USD feed] to get ETH per USDC
+        bool[] shouldReverses;
+    }
+
+    /// @notice Cancel the recurring purchase for the current nonce
+    function cancel() external {
+        // Not explicitly clearing the nonce just cancels the replayable txn
+    }
+
+    /**
+     * @notice Execute a swap given a configuration for a recurring purchase
+     * @param config The configuration for a recurring purchase order
+     */
+    function purchase(PurchaseConfig calldata config) public {
+        allowReplay();
```

Comment:
Hmm, I do remember that about one of the `allowX` methods, but also don't quite remember the exact details. Though I also have a vague recollection it had to do with callbacks, rather than replays. I could look through old PRs to try to dig it up, but I'm glad we're getting rid of `allowReplay()` in v2 so it's no longer relevant in this context. Though, it would still be good to figure out what that exact edge-case is so we could document it in `QuarkScripts.sol`.